### PR TITLE
chore: only add a swap approve transaction if the allowance is not enough

### PR DESCRIPTION
**Pull request by kathaypacific**
### Description

Since gas fees are expensive in some networks, it would be beneficial for users if we could minimise any redundant transactions sent to the network. Currently for the swap flow, we will always create an approval transaction for an ERC-20 sell token. There could possibly be a scenario where a user has approved a larger amount in the past, and the swap amount in question still fits into the allowance.

This PR adds a check for the spender allowance, before adding the approval transaction to the list.

### Test plan

Verified that approval transactions are omitted if there is enough allowance - this is pretty edge case for a user who is swapping on Valora only since we only approve the necessary amount. I tested by hacking an approval that is very large, and checking the subsequent swap does not contain an approval transaction.

I still think there's value here though, e.g.
1. an approval succeeds but the swap transaction fails. the user can then try again without having to pay for the approval again.
2. a user is swapping a very large amount, the swap transaction succeeds but due to slippage, some allowance is left over (enough to cover a subsequent swap that is small)

### Related issues

- Fixes RET-942

### Backwards compatibility

Y

### CODE REVIEW

File: src/swap/useSwapQuote.ts [hunk 1]
```diff
@@ -90,17 +103,18 @@ function createBaseSwapTransactions(
   }
 }
 
-export async function prepareSwapTransactions(
+async function prepareSwapTransactions(
   fromToken: TokenBalance,
   updatedField: Field,
   unvalidatedSwapTransaction: SwapTransaction,
-  price: string,
```
Comment (kathaypacific) [inline src/swap/useSwapQuote.ts:97]: price is unused so i removed it

File: src/swap/useSwapQuote.ts [hunk 2]
```diff
@@ -54,21 +56,32 @@ function createBaseSwapTransactions(
       ? BigInt(new BigNumber(buyAmount).times(guaranteedPrice).toFixed(0, 0))
       : BigInt(sellAmount)
 
-  // Approve transaction if the sell token is ERC-20
-  // TODO skip this if the allowance is already enough
-  if (allowanceTarget !== zeroAddress && fromToken.address) {
-    const data = encodeFunctionData({
+  // If the sell token is ERC-20, we need to check the allowance and add an
+  // approval transaction if necessary
+  if (allowanceTarget !== zeroAddress && fromToken.address && isAddress(fromToken.address)) {
```
Comment (jeanregisser) [inline src/swap/useSwapQuote.ts:61]: nit: Why checking `isAddress(fromToken.address)`? seems a bit too defensive, as if we were not trusting our own data.

File: src/swap/useSwapQuote.ts [hunk 3]
```diff
@@ -126,6 +140,12 @@ function useSwapQuote(networkId: NetworkId, slippagePercentage: string) {
       swapAmount: ParsedSwapAmount,
       updatedField: Field
     ) => {
+      if (!walletAddress || !isAddress(walletAddress)) {
```
Comment (jeanregisser) [inline src/swap/useSwapQuote.ts:143]: nit: we usually aren't that defensive about walletAddress.
```suggestion
      if (!walletAddress) {
```

**Review by jeanregisser — APPROVED** [review]
This is great!! 🙌

File: src/swap/SwapScreen.tsx [hunk 4]
```diff
@@ -108,8 +90,6 @@ export function SwapScreen({ route }: Props) {
   ).maxSlippagePercentage
   const parsedSlippagePercentage = new BigNumber(slippagePercentage).toFormat()
 
-  const useViemForSwap = getFeatureGate(StatsigFeatureGates.USE_VIEM_FOR_SWAP)
```
Comment (dievazqu) [inline src/swap/SwapScreen.tsx:111]: So this PR also forces the new client to use `viem`.
I think this is fine, but I didn't expect this by reading the PR title or description.

**Review by dievazqu — APPROVED** [review]
LGTM

File: src/swap/SwapScreen.tsx (hunk 4 above)
Comment (jeanregisser) [reply src/swap/SwapScreen.tsx:111]: It's because it was initially based on my branch where I removed ContractKit, but then this PR got switched to the main branch when it was merged:

Screenshot 2023-12-15 at 10 15 39

File: src/swap/useSwapQuote.ts (hunk 2 above)
Comment (kathaypacific) [reply src/swap/useSwapQuote.ts:61]: the viem functions require the address to be `Address` type, so typescript complains about the address being the `string` type (less specific). i guess broadly we've been using `as Address`, but we discussed recently to avoid type casting so i went with the typeguard here. i think it should make no difference at runtime. i'm not sure what is best, going for consistency + type casting or going with the type guard (which IMO is more correct)?

File: src/swap/useSwapQuote.ts (hunk 3 above)
Comment (kathaypacific) [reply src/swap/useSwapQuote.ts:143]: same reason as above...walletAddress has type string which means later I need to type cast it to `Address` to use with viem...perhaps i should just change the type of walletAddress to be `Address` in redux, but that'll be a bigger change...

File: src/swap/SwapScreen.tsx (hunk 4 above)
Comment (kathaypacific) [reply src/swap/SwapScreen.tsx]: yes sorry i should have mentioned this 😅 i decided not to bother with the contractkit flow because Jean's PR was about to be merged

File: src/swap/useSwapQuote.ts (hunk 2 above)
Comment (kathaypacific) [reply src/swap/useSwapQuote.ts]: i think consistency is top priority and makes it easier to do a project wide refactor if we want in the future...so i've removed the type guards and did `as Address` in the viem function directly.

Comment (jeanregisser) [reply src/swap/useSwapQuote.ts]: I think we should turn out tokensInfo address field (and others) to be `Address` this would remain compatible with things not using viem and will avoid the need to casting when using viem.

What do you think?

Comment (kathaypacific) [reply src/swap/useSwapQuote.ts]: yes i agree - we should update the redux type to be Address, i can create a task so that we don't forget
